pkg/morserino_com/list_com_test.go

```diff
@@ -0,0 +1,299 @@
+package morserino_com
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.bug.st/serial/enumerator"
+)
+
+//
+// ============ Mock definitions ==================
+
+//Happy case mock
+type mockEnumeratePorts_OK struct{}
+
+//Happy case mocked method
+func (e mockEnumeratePorts_OK) GetDetailedPortsList() ([]*enumerator.PortDetails, error) {
+	happyCaseList := []*enumerator.PortDetails{}
+
+	item1 := enumerator.PortDetails{
+		Name:         "COM Port Name",
+		IsUSB:        true,
+		VID:          "XXXX",
+		PID:          "YYYY",
+		SerialNumber: "001",
+	}
+	item2 := enumerator.PortDetails{
+		Name:         "Morserino Port",
+		IsUSB:        true,
+		VID:          "10C4",
+		PID:          "EA60",
+		SerialNumber: "002",
+	}
+	happyCaseList = append(happyCaseList, &item1)
+	happyCaseList = append(happyCaseList, &item2)
+
+	return happyCaseList, nil
+}
+
+// Mock that returns an error
+type mockEnumeratePorts_error struct{}
+
+// Mocked method that returns an error
+func (e mockEnumeratePorts_error) GetDetailedPortsList() ([]*enumerator.PortDetails, error) {
+	return nil, fmt.Errorf("An error occured")
+}
+
+// Mock that found no ports
+type mockEnumeratePorts_noPort struct{}
+
+// Mocked method that found no ports
+func (e mockEnumeratePorts_noPort) GetDetailedPortsList() ([]*enumerator.PortDetails, error) {
+	emptyList := []*enumerator.PortDetails{}
+
+	return emptyList, nil
+}
```

Comment:
Even if you approach is perfectly valid I'm going to expose to you another approach.

Here you are actually I implementing one mock per behaviour you want to simulate. What you could do instead is define a parametrized mock, and let the set setup actually configure your mock implementation to simulate your case.

Which would give the following.

```golang
type mockPortEnumerator struct {
        ports []*enumarator.PortDetails
        err error
}

func (e mockPortEnumerator) GetDetailedPortsList() ([]*enumerator.PortDetails, error) {
        return e.ports, e.err
}
```

Then you could replace `mockEnumeratePorts_OK` with

```golang
// ...
portEnumerator := mockPortEnumerator{
    ports: []*enumerator.PortDetails{
        {/* Value 1*/},
        {/* Value 2 */},
    },
}
```

And `mockEnumeratePorts_error` with

```golang
// ...
portEnumerator := mockPortEnumerator{
   err: errors.New("Something went wrong"), 
}
```

Does it makes sense for you ?

Reply:
It keeps the test data nearer to the test. Learned a new trick. Thanks a lot.